### ISSUE TITLE
Streaming incoming file to an outgoing request

### ISSUE THREAD

**Issue by jackwanders** [balderdashy/skipper#119]
I am working with `skipper@0.5.8` via `sails@0.11.3` to manage file uploads. My specific use case calls for taking incoming file uploads and sending them to a third party for storage via a `POST` request.

My initial attempt at writing an adapter involved the following logic (abbreviated code sample, don't mind typos or syntax issues):

```javascript
var request = require('request');

module.exports = function(options) {
  return {
    receive: function(options) {
      var receiver__ = new stream.Writable({objectMode: true});
      
      receiver__.on('error', function(err) {
        console.log('Upload error: ', err);
      });

      receiver__._write = function(__newFile, encoding, done) {
        
        request.post({
          url: 'http://www.mystorage.com/path/to/storage/endpoint',
          formData: {
            file: __newFile
          }
        }, function(err, response, body) {
          if (err) {
            receiver__.emit('error', err);
            return;          
          }
          __newFile.extra = body;
          done();
        });

      };
    }
  };
};
```

My justification here is that `__newFile` is a readable stream when we're within the `receiver__._write` function, so I should be able to take advantage of `request`'s [ability to use streams in `multipart/form-data` requests](https://github.com/request/request#multipartform-data-multipart-form-uploads).

However, this doesn't appear to work. I get a 500 response from my external storage service, and I'm finding it difficult to debug.

I *was* able to get things to work by taking the following actions
- In `receiver__._write`, create a write stream to a file path (via `fs.createWriteStream`)
- Pipe`__newFile` to the writable file stream (this and the above step are copy/paste from `skipper-disk`)
- call `setImmediate` (allowing I/O actions to occur, which will kick off the writing of the file)
  - within the `setImmediate` callback, use `fs.createReadStream` to create a readable stream from the file being written to
  - call `request.post` and send the readable stream

While this "write and read a file at the same time" method works, but seems unnecessary. My guess is that there is something in the `request` library that doesn't work with streams where `objectMode: true` has been set. I was hoping you might be able to shed some light on the implementation of `__newFile` and how I may be able to manipulate it to get a readable stream that acts in a way similar to those created by `fs.createReadStream`.

**Comment by jackwanders** [comment]
After some further debugging in both `skipper` and `request`, I think I've found the root cause behind my issue.

`request` has a helper function for determining if a formData field value is a stream or not.

https://github.com/request/request/blob/v2.67.0/lib/helpers.js#L41

It assumes any stream has `.path` and `.mode` properties. While these are both properties of streams created by `fs.createReadStream`, they are not properties of streams in general.

When I add the following two lines to my adapter, `request` successfully sends the file data contained in `__newFile`:

```javascript
__newFile.path = __newfile.fd;
__newFile.mode = 438; // this is what fs.createReadStream sets on it streams
```

So, it looks like this isn't an issue in `skipper`, but rather an issue in what `request` belives is a valid file stream. If I create an issue at https://github.com/request/request I'll reference it here for documentation's sake.

**Comment by jackwanders** [comment]
I've opened an issue concerning the types of streams supported by `request`:

https://github.com/request/request/issues/1949

**Comment by sailsbot** [comment]
Thanks for posting, @jackwanders.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by danywheeler** [comment]
@jackwanders  I have the same needs, but used your code it's not working .
![image](https://user-images.githubusercontent.com/8910772/34195298-b80b356e-e597-11e7-81e1-13f54b7be52e.png)
this is what send to other storage service, it uncompleted.

![image](https://user-images.githubusercontent.com/8910772/34195340-fd73ca8a-e597-11e7-83f9-a78632b6c1f8.png)
this is what I did.
